### PR TITLE
DT-1051: 🐛 Allow delius users to reset password even if auth account …

### PR DESCRIPTION
**Pull request by petergphillips**
…disabled and locked

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/oauth2server/security/UserService.kt [hunk 1]
```diff
@@ -43,6 +43,13 @@ class UserService(
       .or { azureUserService.getAzureUserByUsername(username).map { UserPersonDetails::class.java.cast(it) } }
       .or { deliusUserService.getDeliusUserByUsername(username).map { UserPersonDetails::class.java.cast(it) } }
 
+  fun findEnabledMasterUserPersonDetails(username: String): UserPersonDetails? =
+    authUserService.getAuthUserByUsername(username).filter { it.isEnabled }.map { UserPersonDetails::class.java.cast(it) }
```
Comment (petergphillips) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/oauth2server/security/UserService.kt:47]: need to also check that the user is enabled in each data source in turn

File: src/main/kotlin/uk/gov/justice/digital/hmpps/oauth2server/verify/ResetPasswordServiceImpl.kt [hunk 2]
```diff
@@ -59,11 +59,12 @@ class ResetPasswordServiceImpl(
       multipleMatchesAndCanBeReset = false
       optionalUser = userRepository.findByUsername(usernameOrEmailAddress.toUpperCase())
         .or {
+          // can't find the user in auth, so look in nomis or delius
           userService.findMasterUserPersonDetails(usernameOrEmailAddress.toUpperCase())
             .flatMap { userPersonDetails: UserPersonDetails ->
               when (AuthSource.fromNullableString(userPersonDetails.authSource)) {
-                AuthSource.nomis -> return@flatMap saveNomisUser(userPersonDetails)
-                AuthSource.delius -> return@flatMap saveDeliusUser(userPersonDetails)
```
Comment (petergphillips) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/oauth2server/verify/ResetPasswordServiceImpl.kt:66]: removed unnecessary `return`

File: src/main/kotlin/uk/gov/justice/digital/hmpps/oauth2server/verify/ResetPasswordServiceImpl.kt [hunk 3]
```diff
@@ -105,17 +106,11 @@ class ResetPasswordServiceImpl(
...
-    } else {
-      val userOptional = userService.findMasterUserPersonDetails(user.username)
-      if (userOptional.isEmpty) {
-        // shouldn't really happen, means that a nomis user exists in auth but not in nomis
-        return TemplateAndParameters(resetUnavailableTemplateId, user.username, user.name)
-      }
-      userOptional.get()
-    }
+
+    val userDetails = userService.findEnabledMasterUserPersonDetails(user.username)
```
Comment (petergphillips) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/oauth2server/verify/ResetPasswordServiceImpl.kt:110]: we don't rely on auth database to tell us if the user is enabled / disabled, instead go to the source

File: src/main/kotlin/uk/gov/justice/digital/hmpps/oauth2server/verify/ResetPasswordServiceImpl.kt [hunk 4]
```diff
@@ -171,8 +166,10 @@ class ResetPasswordServiceImpl(
   override fun setPassword(token: String, password: String?) {
     val userToken = userTokenRepository.findById(token).orElseThrow()
     val user = userToken.user
-    val userPersonDetails =
-      if (user.isMaster) user else userService.findMasterUserPersonDetails(user.username).orElseThrow()
+
+    val userPersonDetails = userService.findEnabledMasterUserPersonDetails(user.username)
```
Comment (petergphillips) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/oauth2server/verify/ResetPasswordServiceImpl.kt:170]: again don't rely on the auth database to tell us where the user is from

File: src/main/kotlin/uk/gov/justice/digital/hmpps/oauth2server/security/UserService.kt [hunk 5]
```diff
@@ -43,6 +43,13 @@ class UserService(
...
+      .or { nomisUserService.getNomisUserByUsername(username).filter { it.isEnabled }.map { UserPersonDetails::class.java.cast(it) } }
```
Comment (JamieDraperUK) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/oauth2server/security/UserService.kt:48]: :heart:

**Review by JamieDraperUK — APPROVED** [review]
LGTM

**Pull request merged**